File: packages/typescript/src/features/secrets/validation.ts
/**
 * Secret Validation Module
 *
 * Provides validation strategies for different types of secrets
 * including API keys, URLs, and custom validation.
 */

import { logger } from "../../logger.ts";
import type {
	CustomValidator,
	ValidationResult,
	ValidationStrategy,
} from "./types.ts";

/**
 * Validation strategy implementations
 */
export const ValidationStrategies: Record<string, CustomValidator> = {
	/**
	 * No validation - always passes
	 */
	none: async (_key: string, _value: string): Promise<ValidationResult> => ({
		isValid: true,
		validatedAt: Date.now(),
	}),

	/**
	 * OpenAI API key validation
	 * Format: sk-... or sk-proj-...
	 */
	"api_key:openai": async (
		_key: string,
		value: string,
	): Promise<ValidationResult> => {
		const validatedAt = Date.now();

		if (!value.startsWith("sk-")) {
			return {
				isValid: false,
				error: 'OpenAI API key must start with "sk-"',
				validatedAt,
			};
		}

		if (value.length < 20) {
			return {
				isValid: false,
				error: "OpenAI API key is too short",
				validatedAt,
			};
		}

		// Optionally verify by making a test request
		const shouldVerify = process.env.VALIDATE_API_KEYS === "true";
		if (shouldVerify) {
			const verified = await verifyOpenAIKey(value);
			if (!verified.isValid) {
				return { ...verified, validatedAt };
			}
		}

		return { isValid: true, validatedAt };
	},

	/**
	 * Anthropic API key validation
	 * Format: sk-ant-...
	 */
	"api_key:anthropic": async (
		_key: string,
		value: string,
	): Promise<ValidationResult> => {
		const validatedAt = Date.now();

		if (!value.startsWith("sk-ant-")) {
			return {
				isValid: false,
				error: 'Anthropic API key must start with "sk-ant-"',
				validatedAt,
			};
		}

		if (value.length < 30) {
			return {
				isValid: false,
				error: "Anthropic API key is too short",
				validatedAt,
			};
		}

		// Optionally verify by making a test request
		const shouldVerify = process.env.VALIDATE_API_KEYS === "true";
		if (shouldVerify) {
			const verified = await verifyAnthropicKey(value);
			if (!verified.isValid) {
				return { ...verified, validatedAt };
			}
		}

		return { isValid: true, validatedAt };
	},

	/**
	 * Groq API key validation
	 * Format: gsk_...
	 */
	"api_key:groq": async (
		_key: string,
		value: string,
	): Promise<ValidationResult> => {
		const validatedAt = Date.now();

		if (!value.startsWith("gsk_")) {
			return {
				isValid: false,
				error: 'Groq API key must start with "gsk_"',
				validatedAt,
			};
		}

		if (value.length < 20) {
			return {
				isValid: false,
				error: "Groq API key is too short",
				validatedAt,
			};
		}

		return { isValid: true, validatedAt };
	},

	/**
	 * Google API key validation
	 * Format: AIza...
	 */
	"api_key:google": async (
		_key: string,
		value: string,
	): Promise<ValidationResult> => {
		const validatedAt = Date.now();

		if (!value.startsWith("AIza")) {
			return {
				isValid: false,
				error: 'Google API key must start with "AIza"',
				validatedAt,
			};
		}

		if (value.length < 30) {
			return {
				isValid: false,
				error: "Google API key is too short",
				validatedAt,
			};
		}

		return { isValid: true, validatedAt };
	},

	/**
	 * Mistral API key validation
	 */
	"api_key:mistral": async (
		_key: string,
		value: string,
	): Promise<ValidationResult> => {
		const validatedAt = Date.now();

		if (value.length < 20) {
			return {
				isValid: false,
				error: "Mistral API key is too short",
				validatedAt,
			};
		}

		return { isValid: true, validatedAt };
	},

	/**
	 * Cohere API key validation
	 */
	"api_key:cohere": async (
		_key: string,
		value: string,
	): Promise<ValidationResult> => {
		const validatedAt = Date.now();

		if (value.length < 20) {
			return {
				isValid: false,
				error: "Cohere API key is too short",
				validatedAt,
			};
		}

		return { isValid: true, validatedAt };
	},

	/**
	 * URL format validation
	 */
	"url:valid": async (
		_key: string,
		value: string,
	): Promise<ValidationResult> => {
		const validatedAt = Date.now();

		try {
			new URL(value);
			return { isValid: true, validatedAt };
		} catch {
			return {
				isValid: false,
				error: "Invalid URL format",
				validatedAt,
			};
		}
	},

	/**
	 * URL reachability validation
	 */
	"url:reachable": async (
		_key: string,
		value: string,
	): Promise<ValidationResult> => {
		const validatedAt = Date.now();

		try {
			new URL(value);
		} catch {
			return {
				isValid: false,
				error: "Invalid URL format",
				validatedAt,
			};
		}

		try {
			const response = await fetch(value, {
				method: "HEAD",
				signal: AbortSignal.timeout(5000),
			});

			if (!response.ok) {
				return {
					isValid: false,
					error: `URL returned status ${response.status}`,
					validatedAt,
				};
			}

			return { isValid: true, validatedAt };
		} catch (error) {
			return {
				isValid: false,
				error: `URL is not reachable: ${error instanceof Error ? error.message : "Unknown error"}`,
				validatedAt,
			};
		}
	},

	/**
	 * Custom validation placeholder
	 */
	custom: async (_key: string, _value: string): Promise<ValidationResult> => {
		// Custom validation should be registered separately
		return {
			isValid: true,
			details: "Custom validation not implemented",
			validatedAt: Date.now(),
		};
	},
};

/**
 * Registry for custom validators
 */
const customValidators: Map<string, CustomValidator> = new Map();

/**
 * Register a custom validator
 */
export function registerValidator(
	name: string,
	validator: CustomValidator,
): void {
	customValidators.set(name, validator);
	logger.debug(`[Validation] Registered custom validator: ${name}`);
}

/**
 * Unregister a custom validator
 */
export function unregisterValidator(name: string): boolean {
	return customValidators.delete(name);
}

/**
 * Get a validator by strategy name
 */
export function getValidator(strategy: string): CustomValidator | undefined {
	// Check built-in strategies first
	if (strategy in ValidationStrategies) {
		return ValidationStrategies[strategy];
	}

	// Check custom validators
	return customValidators.get(strategy);
}

/**
 * Validate a secret value
 */
export async function validateSecret(
	key: string,
	value: string,
	strategy?: string,
): Promise<ValidationResult> {
	const strategyName = strategy ?? "none";
	const validator = getValidator(strategyName);

	if (!validator) {
		logger.warn(`[Validation] Unknown validation strategy: ${strategyName}`);
		return {
			isValid: true,
			details: `Unknown validation strategy: ${strategyName}`,
			validatedAt: Date.now(),
		};
	}

	try {
		return await validator(key, value);
	} catch (error) {
		const errorMessage =
			error instanceof Error ? error.message : "Validation error";
		logger.error(`[Validation] Error validating ${key}: ${errorMessage}`);
		return {
			isValid: false,
			error: errorMessage,
			validatedAt: Date.now(),
		};
	}
}

/**
 * Infer validation strategy from secret key name
 */
export function inferValidationStrategy(key: string): ValidationStrategy {
	const upperKey = key.toUpperCase();

	if (upperKey.includes("OPENAI") && upperKey.includes("KEY")) {
		return "api_key:openai";
	}

	if (upperKey.includes("ANTHROPIC") && upperKey.includes("KEY")) {
		return "api_key:anthropic";
	}

	if (upperKey.includes("GROQ") && upperKey.includes("KEY")) {
		return "api_key:groq";
	}

	if (upperKey.includes("GOOGLE") && upperKey.includes("KEY")) {
		return "api_key:google";
	}

	if (upperKey.includes("MISTRAL") && upperKey.includes("KEY")) {
		return "api_key:mistral";
	}

	if (upperKey.includes("COHERE") && upperKey.includes("KEY")) {
		return "api_key:cohere";
	}

	if (upperKey.includes("URL") || upperKey.includes("ENDPOINT")) {
		return "url:valid";
	}

	return "none";
}

// ============================================================================
// API Key Verification Helpers
// ============================================================================

/**
 * Verify OpenAI API key by making a test request
 */
async function verifyOpenAIKey(
	apiKey: string,
): Promise<{ isValid: boolean; error?: string }> {
	try {
		const response = await fetch("https://api.openai.com/v1/models", {
			method: "GET",
			headers: {
				Authorization: `Bearer ${apiKey}`,
			},
			signal: AbortSignal.timeout(10000),
		});

		if (response.status === 401) {
			return { isValid: false, error: "Invalid API key" };
		}

		if (response.status === 429) {
			// Rate limited but key is valid
			return { isValid: true };
		}

		if (!response.ok) {
			return {
				isValid: false,
				error: `API returned status ${response.status}`,
			};
		}

		return { isValid: true };
	} catch (error) {
		return {
			isValid: false,
			error: `Failed to verify: ${error instanceof Error ? error.message : "Unknown error"}`,
		};
	}
}

/**
 * Verify Anthropic API key by making a test request
 */
async function verifyAnthropicKey(
	apiKey: string,
): Promise<{ isValid: boolean; error?: string }> {
	try {
		const response = await fetch("https://api.anthropic.com/v1/messages", {
			method: "POST",
			headers: {
				"x-api-key": apiKey,
				"anthropic-version": "2023-06-01",
				"content-type": "application/json",
			},
			body: JSON.stringify({
				model: "claude-3-haiku-20240307",
				max_tokens: 1,
				messages: [{ role: "user", content: "Hi" }],
			}),
			signal: AbortSignal.timeout(10000),
		});

		if (response.status === 401) {
			return { isValid: false, error: "Invalid API key" };
		}

		if (response.status === 429) {
			// Rate limited but key is valid
			return { isValid: true };
		}

		// 400 with "model" error means key is valid but request is invalid
		// We don't care about that for validation
		if (response.status === 400) {
			const body = (await response.json().catch(() => ({}))) as {
				error?: { type?: string };
			};
			if (body.error?.type === "invalid_request_error") {
				return { isValid: true };
			}
		}

		if (!response.ok && response.status !== 400) {
			return {
				isValid: false,
				error: `API returned status ${response.status}`,
			};
		}

		return { isValid: true };
	} catch (error) {
		return {
			isValid: false,
			error: `Failed to verify: ${error instanceof Error ? error.message : "Unknown error"}`,
		};
	}
}

// ============================================================================
// Exports
// ============================================================================

export { validateSecret as validate };

File: packages/python/elizaos/features/core_capabilities/secrets/validation.py
"""Secret Validation Module.

Provides validation strategies for different types of secrets
including API keys, URLs, and custom validation.

Ported from secrets/validation.ts.
"""

from __future__ import annotations

import logging
import os
import time
from urllib.parse import urlparse

from .types import CustomValidator, ValidationResult

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Built-in validation strategies
# ---------------------------------------------------------------------------


async def _validate_none(_key: str, _value: str) -> ValidationResult:
    """No validation -- always passes."""
    return ValidationResult(is_valid=True, validated_at=time.time())


async def _validate_openai_key(_key: str, value: str) -> ValidationResult:
    """OpenAI API key validation. Format: sk-... or sk-proj-..."""
    validated_at = time.time()

    if not value.startswith("sk-"):
        return ValidationResult(
            is_valid=False,
            error='OpenAI API key must start with "sk-"',
            validated_at=validated_at,
        )
    if len(value) < 20:
        return ValidationResult(
            is_valid=False,
            error="OpenAI API key is too short",
            validated_at=validated_at,
        )

    if os.environ.get("VALIDATE_API_KEYS") == "true":
        verified = await _verify_openai_key(value)
        if not verified.is_valid:
            return ValidationResult(
                is_valid=False,
                error=verified.error,
                validated_at=validated_at,
            )

    return ValidationResult(is_valid=True, validated_at=validated_at)


async def _validate_anthropic_key(_key: str, value: str) -> ValidationResult:
    """Anthropic API key validation. Format: sk-ant-..."""
    validated_at = time.time()

    if not value.startswith("sk-ant-"):
        return ValidationResult(
            is_valid=False,
            error='Anthropic API key must start with "sk-ant-"',
            validated_at=validated_at,
        )
    if len(value) < 30:
        return ValidationResult(
            is_valid=False,
            error="Anthropic API key is too short",
            validated_at=validated_at,
        )

    if os.environ.get("VALIDATE_API_KEYS") == "true":
        verified = await _verify_anthropic_key(value)
        if not verified.is_valid:
            return ValidationResult(
                is_valid=False,
                error=verified.error,
                validated_at=validated_at,
            )

    return ValidationResult(is_valid=True, validated_at=validated_at)


async def _validate_groq_key(_key: str, value: str) -> ValidationResult:
    """Groq API key validation. Format: gsk_..."""
    validated_at = time.time()

    if not value.startswith("gsk_"):
        return ValidationResult(
            is_valid=False,
            error='Groq API key must start with "gsk_"',
            validated_at=validated_at,
        )
    if len(value) < 20:
        return ValidationResult(
            is_valid=False,
            error="Groq API key is too short",
            validated_at=validated_at,
        )

    return ValidationResult(is_valid=True, validated_at=validated_at)


async def _validate_google_key(_key: str, value: str) -> ValidationResult:
    """Google API key validation. Format: AIza..."""
    validated_at = time.time()

    if not value.startswith("AIza"):
        return ValidationResult(
            is_valid=False,
            error='Google API key must start with "AIza"',
            validated_at=validated_at,
        )
    if len(value) < 30:
        return ValidationResult(
            is_valid=False,
            error="Google API key is too short",
            validated_at=validated_at,
        )

    return ValidationResult(is_valid=True, validated_at=validated_at)


async def _validate_mistral_key(_key: str, value: str) -> ValidationResult:
    """Mistral API key validation."""
    validated_at = time.time()
    if len(value) < 20:
        return ValidationResult(
            is_valid=False,
            error="Mistral API key is too short",
            validated_at=validated_at,
        )
    return ValidationResult(is_valid=True, validated_at=validated_at)


async def _validate_cohere_key(_key: str, value: str) -> ValidationResult:
    """Cohere API key validation."""
    validated_at = time.time()
    if len(value) < 20:
        return ValidationResult(
            is_valid=False,
            error="Cohere API key is too short",
            validated_at=validated_at,
        )
    return ValidationResult(is_valid=True, validated_at=validated_at)


async def _validate_url_valid(_key: str, value: str) -> ValidationResult:
    """URL format validation."""
    validated_at = time.time()
    try:
        result = urlparse(value)
        if not result.scheme or not result.netloc:
            raise ValueError("Missing scheme or netloc")
        return ValidationResult(is_valid=True, validated_at=validated_at)
    except Exception:
        return ValidationResult(
            is_valid=False,
            error="Invalid URL format",
            validated_at=validated_at,
        )


async def _validate_url_reachable(_key: str, value: str) -> ValidationResult:
    """URL reachability validation."""
    validated_at = time.time()

    try:
        result = urlparse(value)
        if not result.scheme or not result.netloc:
            raise ValueError("Missing scheme or netloc")
    except Exception:
        return ValidationResult(
            is_valid=False,
            error="Invalid URL format",
            validated_at=validated_at,
        )

    try:
        import aiohttp

        async with (
            aiohttp.ClientSession() as session,
            session.head(
                value,
                timeout=aiohttp.ClientTimeout(total=5),
            ) as resp,
        ):
            if resp.status >= 400:
                return ValidationResult(
                    is_valid=False,
                    error=f"URL returned status {resp.status}",
                    validated_at=validated_at,
                )
            return ValidationResult(is_valid=True, validated_at=validated_at)
    except ImportError:
        # aiohttp not available, fall back to format check only
        return ValidationResult(is_valid=True, validated_at=validated_at)
    except Exception as exc:
        return ValidationResult(
            is_valid=False,
            error=f"URL is not reachable: {exc}",
            validated_at=validated_at,
        )


async def _validate_custom(_key: str, _value: str) -> ValidationResult:
    """Custom validation placeholder."""
    return ValidationResult(
        is_valid=True,
        details="Custom validation not implemented",
        validated_at=time.time(),
    )


# ---------------------------------------------------------------------------
# Strategy registry
# ---------------------------------------------------------------------------

VALIDATION_STRATEGIES: dict[str, CustomValidator] = {
    "none": _validate_none,
    "api_key:openai": _validate_openai_key,
    "openai": _validate_openai_key,
    "api_key:anthropic": _validate_anthropic_key,
    "anthropic": _validate_anthropic_key,
    "api_key:groq": _validate_groq_key,
    "groq": _validate_groq_key,
    "api_key:google": _validate_google_key,
    "google": _validate_google_key,
    "api_key:mistral": _validate_mistral_key,
    "api_key:cohere": _validate_cohere_key,
    "url:valid": _validate_url_valid,
    "url:reachable": _validate_url_reachable,
    "custom": _validate_custom,
}

_custom_validators: dict[str, CustomValidator] = {}


def register_validator(name: str, validator: CustomValidator) -> None:
    """Register a custom validator."""
    _custom_validators[name] = validator
    logger.debug("[Validation] Registered custom validator: %s", name)


def unregister_validator(name: str) -> bool:
    """Unregister a custom validator."""
    return _custom_validators.pop(name, None) is not None


def get_validator(strategy: str) -> CustomValidator | None:
    """Get a validator by strategy name."""
    if strategy in VALIDATION_STRATEGIES:
        return VALIDATION_STRATEGIES[strategy]
    return _custom_validators.get(strategy)


async def validate_secret(
    key: str,
    value: str,
    strategy: str | None = None,
) -> ValidationResult:
    """Validate a secret value."""
    strategy_name = strategy or "none"
    validator = get_validator(strategy_name)

    if not validator:
        logger.warning("[Validation] Unknown validation strategy: %s", strategy_name)
        return ValidationResult(
            is_valid=True,
            details=f"Unknown validation strategy: {strategy_name}",
            validated_at=time.time(),
        )

    try:
        return await validator(key, value)
    except Exception as exc:
        error_message = str(exc)
        logger.error("[Validation] Error validating %s: %s", key, error_message)
        return ValidationResult(
            is_valid=False,
            error=error_message,
            validated_at=time.time(),
        )


def infer_validation_strategy(key: str) -> str:
    """Infer validation strategy from secret key name."""
    upper_key = key.upper()

    if "OPENAI" in upper_key and "KEY" in upper_key:
        return "api_key:openai"
    if "ANTHROPIC" in upper_key and "KEY" in upper_key:
        return "api_key:anthropic"
    if "GROQ" in upper_key and "KEY" in upper_key:
        return "api_key:groq"
    if "GOOGLE" in upper_key and "KEY" in upper_key:
        return "api_key:google"
    if "MISTRAL" in upper_key and "KEY" in upper_key:
        return "api_key:mistral"
    if "COHERE" in upper_key and "KEY" in upper_key:
        return "api_key:cohere"
    if "URL" in upper_key or "ENDPOINT" in upper_key:
        return "url:valid"
    return "none"


# ---------------------------------------------------------------------------
# API key verification helpers
# ---------------------------------------------------------------------------


async def _verify_openai_key(api_key: str) -> ValidationResult:
    """Verify OpenAI API key by making a test request."""
    try:
        import aiohttp

        async with (
            aiohttp.ClientSession() as session,
            session.get(
                "https://api.openai.com/v1/models",
                headers={"Authorization": f"Bearer {api_key}"},
                timeout=aiohttp.ClientTimeout(total=10),
            ) as resp,
        ):
            if resp.status == 401:
                return ValidationResult(
                    is_valid=False, error="Invalid API key", validated_at=time.time()
                )
            if resp.status == 429:
                return ValidationResult(is_valid=True, validated_at=time.time())
            if resp.status >= 400:
                return ValidationResult(
                    is_valid=False,
                    error=f"API returned status {resp.status}",
                    validated_at=time.time(),
                )
            return ValidationResult(is_valid=True, validated_at=time.time())
    except ImportError:
        return ValidationResult(is_valid=True, validated_at=time.time())
    except Exception as exc:
        return ValidationResult(
            is_valid=False, error=f"Failed to verify: {exc}", validated_at=time.time()
        )


async def _verify_anthropic_key(api_key: str) -> ValidationResult:
    """Verify Anthropic API key by making a test request."""
    try:
        import aiohttp

        async with (
            aiohttp.ClientSession() as session,
            session.post(
                "https://api.anthropic.com/v1/messages",
                headers={
                    "x-api-key": api_key,
                    "anthropic-version": "2023-06-01",
                    "content-type": "application/json",
                },
                json={
                    "model": "claude-3-haiku-20240307",
                    "max_tokens": 1,
                    "messages": [{"role": "user", "content": "Hi"}],
                },
                timeout=aiohttp.ClientTimeout(total=10),
            ) as resp,
        ):
            if resp.status == 401:
                return ValidationResult(
                    is_valid=False, error="Invalid API key", validated_at=time.time()
                )
            if resp.status == 429:
                return ValidationResult(is_valid=True, validated_at=time.time())
            if resp.status == 400:
                body = await resp.json()
                if body.get("error", {}).get("type") == "invalid_request_error":
                    return ValidationResult(is_valid=True, validated_at=time.time())
            if resp.status >= 400 and resp.status != 400:
                return ValidationResult(
                    is_valid=False,
                    error=f"API returned status {resp.status}",
                    validated_at=time.time(),
                )
            return ValidationResult(is_valid=True, validated_at=time.time())
    except ImportError:
        return ValidationResult(is_valid=True, validated_at=time.time())
    except Exception as exc:
        return ValidationResult(
            is_valid=False, error=f"Failed to verify: {exc}", validated_at=time.time()
        )


# Alias for compatibility
validate = validate_secret
